airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/format/java/commands.py

```diff
@@ -0,0 +1,75 @@
+import logging
+import sys
+
+import asyncclick as click
+import dagger
+from pipelines.airbyte_ci.format.consts import DEFAULT_FORMAT_IGNORE_LIST
+from pipelines.helpers.utils import sh_dash_c
+
+
+@click.command()
+@click.pass_context
+async def java(ctx: click.Context):
+    """Format java, groovy, and sql code via spotless."""
+    fix = ctx.obj.get("fix_formatting")
+    if fix is None:
+        raise click.UsageError("You must specify either --fix or --check")
+
+    success = await format_java(fix)
+    if not success:
+        click.Abort()
+
+
+async def format_java(fix: bool) -> bool:
+    logger = logging.getLogger("format")
+    if fix:
+        gradle_command = ["./gradlew", "spotlessApply"]
+    else:
+        gradle_command = ["./gradlew", "spotlessCheck", "--scan"]
+
+    async with dagger.Connection(dagger.Config(log_output=sys.stderr)) as dagger_client:
+        try:
+            format_container = await (
+                dagger_client.container()
+                .from_("openjdk:17.0.1-jdk-slim")
+                .with_exec(
+                    sh_dash_c(
+                        [
+                            "apt-get update",
+                            "apt-get install -y bash",
+                        ]
+                    )
+                )
+                .with_mounted_directory(
+                    "/src",
+                    dagger_client.host().directory(
+                        ".",
+                        include=[
+                            "**/*.java",
+                            "**/*.sql",
+                            "**/*.gradle",
+                            "gradlew",
+                            "gradlew.bat",
+                            "gradle",
+                            "**/deps.toml",
+                            "**/gradle.properties",
+                            "**/version.properties",
+                            "tools/gradle/codestyle/java-google-style.xml",
+                            "tools/gradle/codestyle/sql-dbeaver.properties",
+                        ],
+                        exclude=DEFAULT_FORMAT_IGNORE_LIST,
+                    ),
+                )
+                .with_workdir(f"/src")
+                .with_exec(gradle_command)
+            )
+
+            await format_container
+            if fix:
+                await format_container.directory("/src").export(".")
+
+            return True
+        except dagger.ExecError as e:
+            logger.error("Format failed")
+            logger.error(e.stderr)
+            return False
```

Comment:
👋 This is what I meant in your demo when I asked "are you using the[ Python GradleTask class](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/gradle.py#L18)" (I know it's utterly confusing when said orally 🤣 ).
Re-using this class (or just its methods) would make sure we have a standard way of running the gradle commands in dagger.  We recently improved it to make sure subsequent run benefit from caching.

I realize that with Lazy subcommand we're stepping away from the `Step` and pipeline paradigm which happen in the `connector` subgroup.
I find this new approach elegant, nevertheless, I'd appreciate code reuse.
Feel free to split this class up to not make all its internal logic depend on the `Step` one and pick the things you need from it.

Reply:
Oh this one! true! Great point, I'll take a look :)

Reply:
I gave this a try - it was really hard without knowing what's really going on in here. The current code although its a "gradle task" command is highly configured to handle gradle commands in connector containers. 

I personally would feel uncomfortable messing around with this too much myself. If you want to stack off of this to split up the gradleTask into a gradle base container in which i can run gradle commands (not just connector gradle commands), that'd be very welcome! 

Here's the commit where I took a crack; it was also hard with the PipelineContext etc,:
https://github.com/airbytehq/airbyte/pull/31831/commits/a22a14b7b57d15bad129e5acbebd5fd5783bb848

Reply:
This is a sign that some further refactoring is required for this change to land properly. I'm not surprised that the PipelineContext and friends get in the way, and that they would need to be changed at some point. Perhaps the time has come, though that would be outside the scope of this PR.